File: backend/application/ports/video_repository.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime


@dataclass(frozen=True, slots=True)
class VideoRenditionView:
    height: int
    object_key: str
    content_type: str
    size: int


@dataclass(frozen=True, slots=True)
class VideoView:
    id: int
    media_id: int
    status: str
    error_message: str | None
    renditions: tuple[VideoRenditionView, ...]
    original_filename: str = ""
    created_at: datetime | None = None
    duration: float | None = None
    title: str = "Видео"
    owner_id: int | None = None
    owner_name: str = "пользователя"
    views_count: int = 0
    likes_count: int = 0
    is_liked_by_current: bool = False
    is_bookmarked_by_current: bool = False
    is_favorited_by_current: bool = False


@dataclass(frozen=True, slots=True)
class VideoRenditionInput:
    height: int
    object_key: str
    content_type: str = "video/mp4"
    size: int = 0
    width: int | None = None
    duration: float | None = None


@dataclass(frozen=True, slots=True)
class VideoAlbumView:
    id: int
    title: str
    videos: tuple[VideoView, ...]


class VideoRepository(ABC):
    @abstractmethod
    async def get_owner_name(self, *, user_id: int) -> str | None: ...

    @abstractmethod
    async def list_albums(
        self,
        *,
        user_id: int,
        viewer_id: int | None = None,
        tab: str = "uploaded",
        search: str = "",
    ) -> tuple[VideoAlbumView, ...]: ...

    @abstractmethod
    async def create_album(self, *, user_id: int, title: str) -> int: ...

    @abstractmethod
    async def album_belongs_to_user(self, *, album_id: int, user_id: int) -> bool: ...

    @abstractmethod
    async def delete_video(self, *, video_id: int, user_id: int) -> int | None: ...

    @abstractmethod
    async def delete_album(self, *, album_id: int, user_id: int) -> tuple[int, ...] | None: ...

    @abstractmethod
    async def source_object_key(self, *, media_id: int) -> str: ...

    @abstractmethod
    async def create(self, *, media_id: int, album_id: int | None = None, title: str = "Видео") -> int: ...

    @abstractmethod
    async def get_for_user(self, *, video_id: int, user_id: int | None = None) -> VideoView | None: ...

    @abstractmethod
    async def record_view(self, *, video_id: int, user_id: int | None = None) -> int | None: ...

    @abstractmethod
    async def toggle_like(self, *, video_id: int, user_id: int) -> tuple[int, bool] | None: ...

    @abstractmethod
    async def add_bookmark(self, *, video_id: int, user_id: int) -> bool | None: ...

    @abstractmethod
    async def remove_bookmark(self, *, video_id: int, user_id: int) -> bool | None: ...

    @abstractmethod
    async def add_favorite(self, *, video_id: int, user_id: int) -> bool | None: ...

    @abstractmethod
    async def remove_favorite(self, *, video_id: int, user_id: int) -> bool | None: ...

    @abstractmethod
    async def complete_transcoding(
        self,
        *,
        video_id: int,
        status: str,
        renditions: tuple[VideoRenditionInput, ...] = (),
        duration: float | None = None,
        error_message: str | None = None,
    ) -> bool: ...

File: services/media/internal/ports/video_repository.go
package ports

import (
	"context"
	"errors"
	"time"
	"uuid"

	"general-project/media/internal/domain"
)

var (
	// ErrForbidden означает, что пользователь не является владельцем ресурса.
	ErrForbidden = errors.New("video management forbidden")
)

// VideoRepository предоставляет media-сервису доступ к video-метаданным.
type VideoRepository interface {
	Create(ctx context.Context, video domain.VideoAsset) (domain.VideoAsset, error)
	FindByID(ctx context.Context, videoID uuid.UUID) (domain.VideoAsset, error)
	FindDetails(ctx context.Context, videoID uuid.UUID, viewerID *uuid.UUID) (VideoDetails, error)
	ListRenditions(ctx context.Context, videoID uuid.UUID) ([]domain.VideoRendition, error)
	ListAlbums(ctx context.Context, ownerID uuid.UUID, viewerID uuid.UUID, tab string, search string) ([]VideoAlbumView, error)
	CreateAlbum(ctx context.Context, album domain.VideoAlbum) (domain.VideoAlbum, error)
	AlbumBelongsToUser(ctx context.Context, albumID uuid.UUID, userID uuid.UUID) (bool, error)
	DeleteAlbum(ctx context.Context, albumID uuid.UUID, userID uuid.UUID) ([]VideoDeletion, error)
	Delete(ctx context.Context, videoID uuid.UUID) error
	MarkProcessing(ctx context.Context, videoID uuid.UUID) error
	Complete(ctx context.Context, videoID uuid.UUID, status string, duration float64, renditions []domain.VideoRendition, updatedAt time.Time) error
	RecordView(ctx context.Context, videoID uuid.UUID, userID *uuid.UUID) (int, error)
	RecordViewWithOutbox(ctx context.Context, videoID uuid.UUID, userID *uuid.UUID, event OutboxEvent) (int, error)
	ToggleLike(ctx context.Context, videoID uuid.UUID, userID uuid.UUID) (InteractionResult, error)
	SetBookmark(ctx context.Context, videoID uuid.UUID, userID uuid.UUID, value bool) (bool, error)
	SetFavorite(ctx context.Context, videoID uuid.UUID, userID uuid.UUID, value bool) (bool, error)
}

// VideoDetails содержит read model видео и пользовательские признаки взаимодействия.
type VideoDetails struct {
	VideoListItem
	Renditions []domain.VideoRendition
}

// VideoListItem содержит отображаемые поля видео для списков и карточек.
type VideoListItem struct {
	ID                 uuid.UUID
	MediaID            uuid.UUID
	AlbumID            *uuid.UUID
	OwnerID            uuid.UUID
	Title              string
	OriginalFilename   string
	Status             string
	Duration           *float64
	CreatedAt          time.Time
	UpdatedAt          time.Time
	ViewsCount         int
	LikesCount         int
	LikedByViewer      bool
	BookmarkedByViewer bool
	FavoritedByViewer  bool
}

// VideoAlbumView содержит альбом и видео, сгруппированные для выдачи списком.
type VideoAlbumView struct {
	ID     *uuid.UUID
	Title  string
	Videos []VideoListItem
}

// VideoDeletion содержит идентификаторы видео и media для удаления альбома.
type VideoDeletion struct {
	VideoID uuid.UUID
	MediaID uuid.UUID
}

// InteractionResult содержит новый статус like и актуальное количество likes.
type InteractionResult struct {
	Liked      bool
	LikesCount int
}

// TranscodePublisher публикует команду для асинхронной обработки видео.
type TranscodePublisher interface {
	Publish(ctx context.Context, request TranscodeRequest) error
}

// TranscodeRequest содержит данные, необходимые video-worker для обработки.
type TranscodeRequest struct {
	VideoID          uuid.UUID `json:"video_id"`
	MediaID          uuid.UUID `json:"media_id"`
	Bucket           string    `json:"bucket"`
	ObjectKey        string    `json:"object_key"`
	RequestedHeights []int     `json:"requested_heights"`
}

// TranscodeCompletion содержит результат работы video-worker.
type TranscodeCompletion struct {
	VideoID    uuid.UUID            `json:"video_id"`
	Status     string               `json:"status"`
	Duration   float64              `json:"duration"`
	Renditions []TranscodeRendition `json:"renditions"`
}

// TranscodeRendition содержит результат обработки одного разрешения.
type TranscodeRendition struct {
	Height      int     `json:"height"`
	ObjectKey   string  `json:"object_key"`
	ContentType string  `json:"content_type"`
	Size        int64   `json:"size"`
	Duration    float64 `json:"duration"`
}
